binding/binding.go

```diff
@@ -101,6 +101,83 @@ func (b *Binding) MatchTopic(exchange string, routingKey string) bool {
 	return b.Exchange == exchange && b.regexp.MatchString(routingKey)
 }
 
+// MatchHeader checks whether the message can be routed on `b` for a
+// header exchange type.
+func (b *Binding) MatchHeader(exchange string, headers *amqp.Table) bool {
+	if b.Exchange != exchange {
+		return false
+	}
+
+	if b.Arguments == nil {
+		if headers == nil {
+			return true
+		}
+		return false
+	}
+
+	if headers == nil {
+		return false
+	}
+
+	bindingArgTable := map[string]interface{}(*b.Arguments)
+	cliHeaders := map[string]interface{}(*headers)
+
+	// @spec-note AMQP 0.9.1
+	//
+	// Any field starting with 'x-' other than 'x-match' is
+	// reserved for future use and will be ignored.
+	//
+	// * 'all' implies that all the other pairs must match the headers
+	// property of a message for that message to be routed (i.e. and AND match)
+	// * 'any' implies that the message should be routed if any of the
+	// fields in the headers property match one of the fields in the
+	// arguments table (i.e. an OR match)
+	//
+	// We arbitrarily choose `all` as the default if none was provided
+	// at binding time.
+	matchType, ok := bindingArgTable["x-match"]
+	if !ok {
```

Comment:
The value of x-match type must be "all" or "any", nothing more
Code from rabbitmq
```
validate_binding(_X, #binding{args = Args}) ->
    case rabbit_misc:table_lookup(Args, <<"x-match">>) of
        {longstr, <<"all">>} -> ok;
        {longstr, <<"any">>} -> ok;
        {longstr, Other}     -> {error,
                                 {binding_invalid,
                                  "Invalid x-match field value ~p; "
                                  "expected all or any", [Other]}};
        {Type,    Other}     -> {error,
                                 {binding_invalid,
                                  "Invalid x-match field type ~p (value ~p); "
                                  "expected longstr", [Type, Other]}};
        undefined            -> ok %% [0]
    end.
```

Reply:
I understand that no other values than `all` or `any` can be provided at binding-time for the `x-match` argument, though it seems possible that none is provided (I'm not sure what the default case does though, my Erlang isn't on point, sorry).

The AMQP specification does not seem to specify a default behaviour in this case, as such I'm not sure what the right way to handle it should be.

Would you prefer we keep this behaviour and have `all` as default, or leave the route inaccessible if no `x-match` argument was provided at binding-time?

Reply:
Sure, I'll do this